File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/plugins/DefaultAWSClientPlugin.kt
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0.
 */

package software.amazon.smithy.swift.codegen.integration.plugins

import software.amazon.smithy.aws.swift.codegen.AWSSwiftDependency
import software.amazon.smithy.codegen.core.Symbol
import software.amazon.smithy.swift.codegen.SwiftWriter
import software.amazon.smithy.swift.codegen.integration.Plugin
import software.amazon.smithy.swift.codegen.model.buildSymbol

class DefaultAWSClientPlugin : Plugin {
    override val className: Symbol
        get() = buildSymbol {
            this.name = "DefaultAWSClientPlugin"
            this.namespace = AWSSwiftDependency.AWS_CLIENT_RUNTIME.target
            dependency(AWSSwiftDependency.AWS_CLIENT_RUNTIME)
        }

    override val isDefault: Boolean
        get() = true

    override fun customInitialization(writer: SwiftWriter) {
        writer.writeInline("\$L(clientName: self.clientName)", className)
    }
}

File: Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Plugins/DefaultAWSClientPlugin.swift
//
// Copyright Amazon.com Inc. or its affiliates.
// All Rights Reserved.
//
// SPDX-License-Identifier: Apache-2.0
//

import ClientRuntime

public class DefaultAWSClientPlugin: Plugin {
    private var clientName: String

    public init(clientName: String) {
        self.clientName = clientName
    }

    public func configureClient(clientConfiguration: inout ClientConfiguration) async throws {
        if var config = clientConfiguration as? (DefaultClientConfiguration
            & AWSDefaultClientConfiguration
            & AWSRegionClientConfiguration) {
            config.retryStrategyOptions = try AWSClientConfigDefaultsProvider.retryStrategyOptions(
                config.awsRetryMode,
                config.maxAttempts
            )
            clientConfiguration = config as! ClientConfiguration
        }
    }
}
